### DISCUSSION TITLE
Am I not trying to update my password correctly?

### DISCUSSION THREAD

**Discussion by kennethd-uh** [devpi/devpi#1051] [General]
Hi, I'm not sure if this is similar to kgaainer's thread of June 7, but I too am a noob to devpi, and I have set up an organizational devpi server. While trying to introduce it to the team today, I realized my users can't update their passwords.  I've changed my own password using the root account on the server itself, but I am hoping non-root users can use the `devpi` client to manage their own passwords.  Example, running from client machine/dev laptop:
```
(venv3.11) kenneth@x1-uh ~/git/uhiu-devpi $ devpi passwd  --debug --verbose  kenneth
new password for user kenneth:
repeat new password for user kenneth:
/kenneth changing password: ********
PATCH https://devpi.uhiudevs.org/kenneth
403 Forbidden: Access was denied to this resource.

Unauthorized: PyPIView failed permission check
```
The server is running in a docker container behind an nginx proxy based on the example nginx config, and that proxy is behind a letsencrypt proxy, but the docker logs do verify that the request is getting all the way to the `devpi` service itself:
```
devpi-server     | 2024-08-01 18:36:42 - root - INFO - [req901] GET /kenneth
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req901] [Rtx479] in-transaction 479
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req901] [Rtx479] closing transaction at 479
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req901] 200 OK 0.002s serial=479 length=None type=application/json
devpi-proxy      | 172.26.0.3 - - [01/Aug/2024:18:36:42 +0000] "GET /kenneth HTTP/1.1" 200 515 "-" "devpi-client/7.0.2 (py3.11.2; linux)" "172.93.10.194"
certbot-proxy    | 172.93.10.194 - - [01/Aug/2024:18:36:42 +0000] "GET /kenneth HTTP/1.1" 200 290 "-" "devpi-client/7.0.2 (py3.11.2; linux)" "-"
devpi-server     | 2024-08-01 18:36:42 - root - INFO - [req902] PATCH /kenneth
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req902] [Wtx479] in-transaction 479
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req902] [Wtx479] nothing to commit, just closing tx
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req902] [Wtx479] closing transaction at 479
devpi-server     | 2024-08-01 18:36:42 - root - DEBUG - [req902] 403 Forbidden 0.002s serial=479 length=0 type=text/html; charset=UTF-8
devpi-proxy      | 172.26.0.3 - - [01/Aug/2024:18:36:42 +0000] "PATCH /kenneth HTTP/1.1" 403 153 "-" "devpi-client/7.0.2 (py3.11.2; linux)" "172.93.10.194"
certbot-proxy    | 172.93.10.194 - - [01/Aug/2024:18:36:42 +0000] "PATCH /kenneth HTTP/1.1" 403 153 "-" "devpi-client/7.0.2 (py3.11.2; linux)" "-"
```
I've been poring over docs, past issues, and source code, and I am unsure, am I just calling `devpi passwd` wrong?  I'm also not sure about the PATCH request with length=0

Thanks for any ideas!

**Comment by kennethd-uh** [comment]
forgot to include:
```
(venv3.11) kenneth@x1-uh ~/git/uhiu-devpi $ pip freeze | grep devpi 
devpi-client==7.1.0
devpi-client-extensions==0.3.1
devpi-common==4.0.4
```
(edit: upgraded packages to match server -- no change in behavior)

**Comment by kennethd-uh** [comment]
and on the server:
```
azureuser@devpi:~$ docker exec -it devpi-server sh -c ". /app/venv3.12/bin/activate && pip freeze | grep devpi"
devpi-client==7.1.0
devpi-common==4.0.4
devpi-server==6.12.1
devpi-web==4.2.2
```

**Comment by kennethd-uh** [comment]
for `devpi-proxy` container, main `nginx.conf` was created with `docker run --rm nginx:1.27-bookworm cat /etc/nginx/nginx.conf > ./docker-files/etc/nginx/devpi-proxy/nginx.conf`, the only thing i changed was adding "daemon off;"

`/etc/nginx/devpi-proxy/conf.d/devpi-proxy.conf` doesn't look much different from `server/devpi_server/cfg/nginx-devpi.conf.template`, just some hardcoded values & omission of some of the caching tokens, which i am not using:
```
1c1
< %(caching_http)smap $http_x_forwarded_proto $x_scheme {
---
> map $http_x_forwarded_proto $x_scheme {
8,9c8,9
<     server_name %(outside_host)s $hostname "";
<     listen %(outside_port)s;
---
>     server_name localhost $hostname "";
>     listen 8080;
23,24c23,24
<     root %(serverdir)s;
< %(caching_server)s
---
>     root /opt/devpi-server;
> 
34c34
< %(caching_location)stry_files /+files$uri @proxy_to_app;
---
>         try_files /+files$uri @proxy_to_app;
41,42c41,42
<         root %(serverdir)s;
< %(caching_location)stry_files $uri @proxy_to_app;
---
>         root /opt/devpi-server;
>         try_files $uri @proxy_to_app;
52c52,53
< %(caching_proxy)sproxy_pass http://localhost:%(port)s;
---
>         # address devpi container by hostname defined in docker-compose.yaml
>         proxy_pass http://devpi:3141;
```

**Comment by kennethd-uh** [comment]
aha, perhaps it is this `restrict-modify` setting?
```
(venv3.11) kenneth@x1-uh ~/git/uhiu-devpi $ cat ./docker-files/etc/devpi-server/devpi-server.yml 
devpi-server:
  serverdir: /opt/devpi-server
  secretfile: /app/var/run/devpi-server-secret
  logger-cfg: /app/etc/devpi-server/logging.yml
  host: 0.0.0.0
  port: 3141
  restrict-modify: root
```

**Comment by fschulze** [comment]
It is very likely to be caused by ``--restrict-modify``. I will have to look into that. If it is, then we should split that permission from general index modification. I guess this didn't come up yet, because most people who use ``--restrict-modify`` also have external user management with for example ``devpi-ldap``.

**Comment by kennethd-uh** [comment]
Oh interesting, we might be interested in contributing to a `devpi-okta` package in the future, `pluggy` looks easy to work with, I've never seen that before. We're a small team though, priority would be something like, do I feel like trying to tackle it over a weekend

**Comment by fschulze** [comment]
Let me know if you have questions, but devpi-ldap should be a good starting point. If you get totally stuck, you can also hire me via email to mail@pyfidelity.com

**Comment by kennethd-uh** [comment]
As an experiment, I built an image without `restrict-modify` and I was able to update my password from client:
```
(venv3.11) kenneth@x1-uh ~/git/uhiu-devpi (main) $ devpi passwd  --debug --verbose  
new password for user kenneth:
repeat new password for user kenneth:
/kenneth changing password: ********
user modified: kenneth
```

**Comment by kennethd-uh** [comment]
I don't know if it was ever meant to work, but setting it to a list of users similar to `acl_upload` also doesn't work: `restrict-modify: root,kenneth,derek,olin,nola,megan,emily`
